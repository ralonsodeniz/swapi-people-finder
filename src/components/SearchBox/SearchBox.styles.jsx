import styled, { css } from 'styled-components';

import mediaQueryHelper from '../../style/media-queries';

const shrinkLabelStyles = css`
  top: -0.5rem;
  font-size: ${({ theme }) => theme.smallFontSize};
  background-image: none;
  -webkit-text-fill-color: ${({ theme }) => theme.titleColor};

  ${mediaQueryHelper(
    'phone',
    css`
      top: -0.4rem;
    `
  )};
`;

export const SearchBoxContainer = styled.div`
  padding: 0.5rem;
  position: relative;
  align-self: center;
`;

export const SearchBoxInput = styled.input`
  color: ${({ theme }) => theme.titleColor};
  padding: 0.5rem;
  border: 1px solid ${({ theme }) => theme.titleColor};
  display: block;
  border-radius: 5px;
  background-color: ${({ theme }) => theme.primaryColor};
  font-size: ${({ theme }) => theme.defaultFontSize};
  cursor: pointer;

  &:focus {
    outline: none;
  }

  &:focus ~ label {
    ${shrinkLabelStyles}
  }

  ${mediaQueryHelper(
    'phone',
    css`
      padding: 0.35rem;
    `
  )};
`;

export const SearchBoxLabel = styled.label`
  font-size: ${({ theme }) => theme.defaultFontSize};
  position: absolute;
  pointer-events: none;
  left: 1rem;
  top: 17.5px;
  transition: 300ms ease all;
  z-index: 9;

  ${mediaQueryHelper(
    'phone',
    css`
      top: 15px;
    `
  )};

  ${({ shrink }) => (shrink ? shrinkLabelStyles : '')};
`;
