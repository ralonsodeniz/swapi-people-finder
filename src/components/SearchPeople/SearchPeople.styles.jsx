import styled, { css } from 'styled-components';

import mediaQueryHelper from '../../style/media-queries';

export const SearchPeopleContainer = styled.div`
  grid-area: search;
  height: 25rem;
  display: grid;
  grid-template: auto / auto auto;
  grid-template-areas:
    'title searchbox'
    'table table'
    'pagination pagination';
  position: relative;
  background: ${({ theme }) => theme.secondaryColor};
  border-radius: 5px;

  ${mediaQueryHelper(
    'small-desktop',
    css`
      height: 17.5rem;
    `
  )};

  ${mediaQueryHelper(
    'tab-xl',
    css`
      height: 25rem;
    `
  )};

  ${mediaQueryHelper(
    'tab-port',
    css`
      height: 17.5rem;
    `
  )};

  ${mediaQueryHelper(
    'phone',
    css`
      height: 19rem;
      grid-template: auto / auto;
      grid-template-areas:
        'title'
        'searchbox'
        'table'
        'pagination';
    `
  )}
`;

export const TableScroll = styled.div`
  grid-area: table;
  padding: 0 0 0 1rem;
  width: 65rem;
  height: 16.5rem;
  overflow-x: auto;
  overflow-y: auto;

  ${mediaQueryHelper(
    'small-desktop',
    css`
      height: 9.5rem;
      width: 45.15rem;
      font-size: ${({ theme }) => theme.smallFontSize};
    `
  )};

  ${mediaQueryHelper(
    'tab-xl',
    css`
      width: 57rem;
      height: 15.5rem;
    `
  )};

  ${mediaQueryHelper(
    'tab-port',
    css`
      height: 9.5rem;
      width: 40.75rem;
      font-size: ${({ theme }) => theme.smallFontSize};
    `
  )};

  ${mediaQueryHelper(
    'phone',
    css`
      width: 18.3rem;
    `
  )};

  ${mediaQueryHelper(
    'phone-xs',
    css`
      width: 17.85rem;
    `
  )};

  ::-webkit-scrollbar {
    height: 1rem;
    display: inline-block;
  }

  ::-webkit-scrollbar-track {
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
    -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.3);
    -webkit-border-radius: 1rem;
    border-radius: 1rem;
  }

  ::-webkit-scrollbar-thumb {
    -webkit-border-radius: 1rem;
    border-radius: 1rem;
    background: rgba(0, 0, 0, 0.4);
    box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.5);
    -webkit-box-shadow: inset 0 0 6px rgba(0, 0, 0, 0.5);
  }
`;

export const SearchPeopleTable = styled.table`
  border-collapse: collapse;
  width: 64rem;
  background-color: ${({ theme }) => theme.cardColor};

  ${mediaQueryHelper(
    'small-desktop',
    css`
      width: 44.15rem;
    `
  )};

  ${mediaQueryHelper(
    'tab-xl',
    css`
      width: 56rem;
    `
  )};

  ${mediaQueryHelper(
    'tab-port',
    css`
      width: 39.75rem;
    `
  )};

  ${mediaQueryHelper(
    'phone',
    css`
      width: 17.25rem;
    `
  )};

  ${mediaQueryHelper(
    'phone-xs',
    css`
      width: 16.85rem;
    `
  )};
`;

export const TableHeader = styled.thead`
  & th {
    position: sticky;
    top: 0;
    padding: 1rem 0 1rem 1rem;
    text-align: left;
    background-color: ${({ theme }) => theme.titleColor};
    color: ${({ theme }) => theme.primaryColor};
    width: max-content;

    ${mediaQueryHelper(
      'small-desktop',
      css`
        padding: 0.5rem 0 0.5rem 0.5rem;
      `
    )};

    ${mediaQueryHelper(
      'tab-port',
      css`
        padding: 0.5rem 0 0.5rem 0.5rem;
      `
    )};
  }
`;

export const TableBody = styled.tbody`
  & th,
  td {
    width: max-content;
    border: none;
    padding: 0.5rem 1rem 0.5rem 1rem;

    ${mediaQueryHelper(
      'small-desktop',
      css`
        padding: 0.25rem 0.5rem 0.25rem 0.5rem;
      `
    )};

    ${mediaQueryHelper(
      'tab-port',
      css`
        padding: 0.25rem 0.5rem 0.25rem 0.5rem;
      `
    )};
  }

  & tr:nth-child(odd) {
    background-color: ${({ theme }) => theme.altCardColor};
  }

  & tr:hover {
    background-color: ${({ theme }) => theme.tableHoverColor};
  }
`;

export const SearchPeopleSpinnerContainer = styled.div`
  background-color: ${({ theme }) => theme.primaryColorOpacity};
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  z-index: 9999;
`;

export const SearchPeopleTitle = styled.h2`
  grid-area: title;
  ${({ theme }) => theme.titleStyles};
  padding-left: 1rem;
  margin: 1rem auto 0.75rem 0;
  color: ${({ theme }) => theme.titleColor};
`;

export const SearchPeopleSearchBox = styled.div`
  grid-area: searchbox;
  display: flex;
  align-items: center;
  padding-right: 1rem;
  margin: 1rem 0 0.75rem auto;

  ${mediaQueryHelper(
    'phone',
    css`
      margin: 0 auto 0.75rem 0;
      padding: 0 0 0 0.5rem;
    `
  )};
`;

export const SearchPeopleButtonsContainer = styled.div`
  display: flex;
  justify-content: flex-end;
`;

export const TablePaginationContainer = styled.div`
  grid-area: pagination;
  padding-left: 1rem;
  display: flex;
`;
