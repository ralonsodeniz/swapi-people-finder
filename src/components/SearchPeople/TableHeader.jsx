import React from 'react';
import PropTypes from 'prop-types';

const TableHeader = ({ deviceType }) => {
  return deviceType === 'phone' || deviceType === 'phone-xs' ? (
    <tr>
      <th>Name</th>
      <th>Gender</th>
      <th />
    </tr>
  ) : (
    <tr>
      <th>Name</th>
      <th>Gender</th>
      <th>Birth year</th>
      <th>Eye color</th>
      <th />
    </tr>
  );
};

TableHeader.propTypes = {
  deviceType: PropTypes.string.isRequired,
};

export default TableHeader;
