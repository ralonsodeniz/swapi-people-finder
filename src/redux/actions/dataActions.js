/* eslint-disable no-console */
import DATA from '../types/dataTypes';
import { getPageDataFromAPI } from '../helpers/dataHelpers';

export const toggleDataLoading = loadingState => ({
  type: DATA.TOGGLE_DATA_LOADING,
  payload: loadingState,
});

export const fetchApiSuccess = searchData => ({
  type: DATA.FETCH_API_SUCCESS,
  payload: searchData,
});

export const fetchApiFailure = error => ({
  type: DATA.FETCH_API_FAILURE,
  payload: error,
});

export const saveCharacter = character => ({
  type: DATA.SAVE_CHARACTER,
  payload: character,
});

export const removeCharacter = characterName => ({
  type: DATA.REMOVE_CHARACTER,
  payload: characterName,
});

export const selectCharacter = character => ({
  type: DATA.SELECT_CHARACTER,
  payload: character,
});

export const setFilter = filter => ({
  type: DATA.SET_FILTER,
  payload: filter,
});

export const setSeachText = searchText => ({
  type: DATA.SET_SEARCH_TEXT,
  payload: searchText,
});

export const fetchApiStart = url => async dispatch => {
  dispatch(toggleDataLoading(true));
  try {
    const searchData = await getPageDataFromAPI(url);
    dispatch(fetchApiSuccess(searchData));
  } catch (error) {
    dispatch(fetchApiFailure(error.message));
    console.log(error);
  }
};
