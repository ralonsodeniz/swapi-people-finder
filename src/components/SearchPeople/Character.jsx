import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';

import { saveCharacter, selectCharacter } from '../../redux/actions/dataActions';

import CustomButton from '../CustomButton/CustomButton';

import { SearchPeopleButtonsContainer } from './SearchPeople.styles';

const Character = ({ character, viewWidth }) => {
  const dispatch = useDispatch();

  const handleSaveCharacter = useCallback(() => {
    dispatch(saveCharacter(character));
  }, [dispatch, saveCharacter, character]);

  const handleSelectCharacter = useCallback(() => {
    dispatch(selectCharacter(character));
  }, [dispatch, selectCharacter, character]);

  return viewWidth <= 600 ? (
    <tr>
      <td>{character.name}</td>
      <td>{character.gender}</td>
      <td>
        <SearchPeopleButtonsContainer>
          <CustomButton
            type="button"
            variant="default"
            size="small"
            onClick={handleSelectCharacter}
          >
            Details
          </CustomButton>
          <CustomButton type="button" variant="save" size="small" onClick={handleSaveCharacter}>
            Save
          </CustomButton>
        </SearchPeopleButtonsContainer>
      </td>
    </tr>
  ) : (
    <tr>
      <td>{character.name}</td>
      <td>{character.gender}</td>
      <td>{character.birthYear}</td>
      <td>{character.eyeColor}</td>
      <td>
        <SearchPeopleButtonsContainer>
          <CustomButton
            type="button"
            variant="default"
            size="small"
            onClick={handleSelectCharacter}
          >
            Details
          </CustomButton>
          <CustomButton type="button" variant="save" size="small" onClick={handleSaveCharacter}>
            Save
          </CustomButton>
        </SearchPeopleButtonsContainer>
      </td>
    </tr>
  );
};

Character.propTypes = {
  character: PropTypes.shape({
    name: PropTypes.string,
    height: PropTypes.string,
    mass: PropTypes.string,
    hairColor: PropTypes.string,
    skinColor: PropTypes.string,
    eyeColor: PropTypes.string,
    birthYear: PropTypes.string,
    gender: PropTypes.string,
  }).isRequired,
  viewWidth: PropTypes.number.isRequired,
};

export default Character;
