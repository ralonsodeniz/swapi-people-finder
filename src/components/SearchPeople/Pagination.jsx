/* eslint-disable react/no-array-index-key */
import React, { useState, useCallback, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import {
  selectCharacterCount,
  selectNextEndpoint,
  selectPreviousEndpoint,
  selectSearchText,
} from '../../redux/reducers/dataReducer';
import { fetchApiStart } from '../../redux/actions/dataActions';

import CustomButton from '../CustomButton/CustomButton';

const selectPaginationData = createStructuredSelector({
  characterCount: selectCharacterCount,
  nextEndpoint: selectNextEndpoint,
  previousEndpoint: selectPreviousEndpoint,
  searchText: selectSearchText,
});

const Pagination = () => {
  const dispatch = useDispatch();
  const [selectedPage, setSelectedPage] = useState(1);
  const { characterCount, nextEndpoint, previousEndpoint, searchText } = useSelector(
    selectPaginationData
  );
  const pageCount = Math.ceil(characterCount / 10);

  useEffect(() => {
    setSelectedPage(1);
  }, [pageCount, setSelectedPage]);

  const handlePageChange = useCallback(
    (mode, pageNumber) => {
      if (mode === 'up') {
        setSelectedPage(prevState => prevState + 1);
        dispatch(fetchApiStart(nextEndpoint));
      } else if (mode === 'down') {
        setSelectedPage(prevState => prevState - 1);
        dispatch(fetchApiStart(previousEndpoint));
      } else if (selectedPage !== pageNumber) {
        setSelectedPage(pageNumber);
        dispatch(
          fetchApiStart(`https://swapi.dev/api/people/?search=${searchText}&page=${pageNumber}`)
        );
      }
    },
    [
      dispatch,
      setSelectedPage,
      nextEndpoint,
      previousEndpoint,
      selectedPage,
      fetchApiStart,
      searchText,
    ]
  );

  const paginationMarkUp = Array.from({ length: pageCount }).map((page, pageIndex) => (
    <CustomButton
      key={pageIndex}
      name="all"
      collapse
      selected={selectedPage === pageIndex + 1}
      size="small"
      onClick={() => handlePageChange('page', pageIndex + 1)}
    >
      {pageIndex + 1}
    </CustomButton>
  ));

  return (
    <>
      <CustomButton
        collapse
        disabled={selectedPage <= 1}
        size="small"
        onClick={() => handlePageChange('down')}
      >
        {`<`}
      </CustomButton>
      {paginationMarkUp}
      <CustomButton
        collapse
        disabled={selectedPage >= pageCount}
        size="small"
        onClick={() => handlePageChange('up')}
      >
        {`>`}
      </CustomButton>
    </>
  );
};

export default Pagination;
