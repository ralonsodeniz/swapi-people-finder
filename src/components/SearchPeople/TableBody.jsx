import React from 'react';
import PropTypes, { shape } from 'prop-types';

import Character from './Character';

const TableBody = ({ filteredSearchArray, deviceType, searchArray }) => {
  return filteredSearchArray.length ? (
    filteredSearchArray.map(character => (
      <Character character={character} deviceType={deviceType} key={character.name} />
    ))
  ) : (
    <tr>
      <td colSpan={deviceType === 'phone' || deviceType === 'phone-xs' ? 3 : 5}>
        {!searchArray.length
          ? 'No results for this search input'
          : 'All characters in this page are already saved'}
      </td>
    </tr>
  );
};

TableBody.propTypes = {
  filteredSearchArray: PropTypes.arrayOf(
    shape({
      name: PropTypes.string,
      height: PropTypes.string,
      mass: PropTypes.string,
      hairColor: PropTypes.string,
      skinColor: PropTypes.string,
      eyeColor: PropTypes.string,
      birthYear: PropTypes.string,
      gender: PropTypes.string,
    })
  ).isRequired,
  deviceType: PropTypes.string.isRequired,
  searchArray: PropTypes.arrayOf(
    shape({
      name: PropTypes.string,
      height: PropTypes.string,
      mass: PropTypes.string,
      hairColor: PropTypes.string,
      skinColor: PropTypes.string,
      eyeColor: PropTypes.string,
      birthYear: PropTypes.string,
      gender: PropTypes.string,
    })
  ).isRequired,
};

export default TableBody;
