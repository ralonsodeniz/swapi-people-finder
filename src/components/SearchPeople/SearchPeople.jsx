import React, { useState, useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import {
  selectSearchArray,
  selectSavedArray,
  selectLoadingData,
} from '../../redux/reducers/dataReducer';
import { fetchApiStart, setSeachText } from '../../redux/actions/dataActions';
import { getFilteredSearchArray } from '../../helpers/filters';
import useDeviceType from '../../helpers/useDeviceType';

import Spinner from '../Spinner/Spinner';
import SearchBox from '../SearchBox/SearchBox';
import CustomButton from '../CustomButton/CustomButton';
import Pagination from './Pagination';
import TableHeader from './TableHeader';
import TableBody from './TableBody';

import {
  SearchPeopleContainer,
  SearchPeopleSearchBox,
  SearchPeopleSpinnerContainer,
  SearchPeopleTable,
  SearchPeopleTitle,
  TableBodyContainer,
  TableHeaderContainer,
  TableScroll,
  TablePaginationContainer,
} from './SearchPeople.styles';

const selectSearchPeopleData = createStructuredSelector({
  searchArray: selectSearchArray,
  savedArray: selectSavedArray,
  loadingData: selectLoadingData,
});

const SearchPeople = () => {
  const dispatch = useDispatch();
  const [searchField, setSearchField] = useState('');
  const { searchArray, savedArray, loadingData } = useSelector(selectSearchPeopleData);

  const filteredSearchArray = getFilteredSearchArray(searchArray, savedArray);

  const deviceType = useDeviceType();

  const handleChange = event => {
    const { value } = event.target;
    setSearchField(value);
  };

  const handleSearch = useCallback(() => {
    dispatch(fetchApiStart(`https://swapi.co/api/people/?search=${searchField}`));
    dispatch(setSeachText(searchField));
    setSearchField('');
  }, [dispatch, searchField, fetchApiStart, setSearchField]);

  return (
    <SearchPeopleContainer>
      {(loadingData || !deviceType) && (
        <SearchPeopleSpinnerContainer>
          <Spinner />
        </SearchPeopleSpinnerContainer>
      )}
      <SearchPeopleTitle>Search people</SearchPeopleTitle>
      <SearchPeopleSearchBox>
        <SearchBox
          name="searchField"
          value={searchField}
          id="searchField"
          onSearchChange={handleChange}
          label="Search people"
        />
        <CustomButton
          type="button"
          variant="default"
          size={deviceType === 'big-desktop' ? 'medium' : 'small'}
          onClick={handleSearch}
        >
          Search
        </CustomButton>
      </SearchPeopleSearchBox>
      <TableScroll>
        <SearchPeopleTable>
          {deviceType && (
            <>
              <TableHeaderContainer>
                <TableHeader deviceType={deviceType} />
              </TableHeaderContainer>
              <TableBodyContainer>
                <TableBody
                  deviceType={deviceType}
                  filteredSearchArray={filteredSearchArray}
                  searchArray={searchArray}
                />
              </TableBodyContainer>
            </>
          )}
        </SearchPeopleTable>
      </TableScroll>
      <TablePaginationContainer>
        <Pagination />
      </TablePaginationContainer>
    </SearchPeopleContainer>
  );
};

export default SearchPeople;
