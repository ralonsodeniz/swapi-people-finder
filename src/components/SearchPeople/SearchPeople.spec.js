import React from 'react';
import { shallow } from 'enzyme';
import * as Redux from 'react-redux';
import configureStore from 'redux-mock-store';
import thunkMiddleware from 'redux-thunk';

import { dataStore, loadingDataStore } from '../../../__mocks__/dataStateMock';
import * as dataActions from '../../redux/actions/dataActions';

import SearchPeople from './SearchPeople';
import { SearchBoxInput } from '../SearchBox/SearchBox.styles';

describe('<SearchPeople /> tests', () => {
  let wrapper;
  let store;
  let useEffect;
  let fetchApiStart;

  const mockUseEffect = () => {
    useEffect.mockImplementationOnce(f => f());
  };

  beforeEach(() => {
    useEffect = jest.spyOn(React, 'useEffect');
    fetchApiStart = jest.spyOn(dataActions, 'fetchApiStart');
  });

  describe('store with data tests', () => {
    beforeEach(() => {
      global.innerWidth = 500;
      global.dispatchEvent(new Event('resize'));
      mockUseEffect();

      store = configureStore([thunkMiddleware])(dataStore);
      jest.spyOn(Redux, 'useDispatch').mockImplementation(() => store.dispatch);
      jest.spyOn(Redux, 'useSelector').mockImplementation(() => store.getState());
      store.clearActions();
      wrapper = shallow(<SearchPeople store={store} />);
    });

    it('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders proper tr for header with 3 th with Name, Gender and "" for mobile devices', () => {
      const tableHeaders = wrapper.find('th');
      expect(tableHeaders.length).toEqual(3);
      const firstTableHeaderText = tableHeaders.at(0).text();
      expect(firstTableHeaderText).toEqual('Name');
      const secondTableHeaderText = tableHeaders.at(1).text();
      expect(secondTableHeaderText).toEqual('Gender');
      const thirdTableHeaderText = tableHeaders.at(2).text();
      expect(thirdTableHeaderText).toEqual('');
    });

    it('filters searchArray with savedArray and renders 7 Characters', () => {
      const characters = wrapper.find('Character');
      expect(characters.length).toEqual(7);
      const firstCharacter = characters.at(0).prop('character');
      expect(firstCharacter.name).toEqual('C-3PO');
      const secondCharacter = characters.at(1).prop('character');
      expect(secondCharacter.name).toEqual('Darth Vader');
      const thirdCharacter = characters.at(2).prop('character');
      expect(thirdCharacter.name).toEqual('Owen Lars');
      const fourthCharacter = characters.at(3).prop('character');
      expect(fourthCharacter.name).toEqual('Beru Whitesun lars');
      const fifthCharacter = characters.at(4).prop('character');
      expect(fifthCharacter.name).toEqual('R5-D4');
      const sixthCharacter = characters.at(5).prop('character');
      expect(sixthCharacter.name).toEqual('Biggs Darklighter');
      const seventhCharacter = characters.at(6).prop('character');
      expect(seventhCharacter.name).toEqual('Obi-Wan Kenobi');
    });

    it('changes searchField to skywalker, sends the fetch action with proper url when button is clicked and sets searchtext and resets searchfield', () => {
      const searchBox = wrapper.find('SearchBox');
      expect(searchBox.length).toEqual(1);
      const searchBoxInput = searchBox.dive().find(SearchBoxInput);
      expect(searchBoxInput.length).toEqual(1);
      searchBoxInput.simulate('change', { target: { value: 'skywalker' } });
      const searchBoxValue = wrapper.find('SearchBox').prop('value');
      expect(searchBoxValue).toEqual('skywalker');
      const button = wrapper.find('CustomButton');
      expect(button.length).toEqual(1);
      button.simulate('click');
      expect(fetchApiStart).toHaveBeenCalledWith(
        `https://swapi.co/api/people/?search=${searchBoxValue}`
      );
      const newSearchBoxValue = wrapper.find('SearchBox').prop('value');
      expect(newSearchBoxValue).toEqual('');
      const actions = store.getActions();
      const expectedActionToogleDataLoading = dataActions.toggleDataLoading(true);
      const expectedActionSetSearchText = dataActions.setSeachText(searchBoxValue);
      expect(actions[0]).toEqual(expectedActionToogleDataLoading);
      expect(actions[1]).toEqual(expectedActionSetSearchText);
    });
  });

  describe('data loading tests', () => {
    beforeEach(() => {
      global.innerWidth = 1920;
      global.dispatchEvent(new Event('resize'));
      mockUseEffect();

      store = configureStore([thunkMiddleware])(loadingDataStore);
      jest.spyOn(Redux, 'useDispatch').mockImplementation(() => store.dispatch);
      jest.spyOn(Redux, 'useSelector').mockImplementation(() => store.getState());
      store.clearActions();
      wrapper = shallow(<SearchPeople store={store} />);
    });

    it('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('renders proper tr for header with 5 th with Name, Gender, Birth year, Eye color and "" for desktop and table devices', () => {
      const tableHeaders = wrapper.find('th');
      expect(tableHeaders.length).toEqual(5);
      const firstTableHeaderText = tableHeaders.at(0).text();
      expect(firstTableHeaderText).toEqual('Name');
      const secondTableHeaderText = tableHeaders.at(1).text();
      expect(secondTableHeaderText).toEqual('Gender');
      const thirdTableHeaderText = tableHeaders.at(2).text();
      expect(thirdTableHeaderText).toEqual('Birth year');
      const fourthTableHeaderText = tableHeaders.at(3).text();
      expect(fourthTableHeaderText).toEqual('Eye color');
      const fifthTableHeaderText = tableHeaders.at(4).text();
      expect(fifthTableHeaderText).toEqual('');
    });

    it('renders Spinner and no Characters but a placeholder td with a colSpan of 5', () => {
      const spinner = wrapper.find('Spinner');
      expect(spinner.length).toEqual(1);
      const characters = wrapper.find('Character');
      expect(characters.length).toEqual(0);
      const placeholder = wrapper.find('td');
      expect(placeholder.length).toEqual(1);
      const placeholderColSpan = placeholder.prop('colSpan');
      expect(placeholderColSpan).toEqual(5);
      const placeholderText = placeholder.text();
      expect(placeholderText).toEqual('All characeters in this page are already saved');
    });
  });
});
