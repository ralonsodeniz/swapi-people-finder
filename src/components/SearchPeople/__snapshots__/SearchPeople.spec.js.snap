// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchPeople /> tests data loading tests matches snapshot 1`] = `
<styled.div>
  <styled.div>
    <Spinner />
  </styled.div>
  <styled.h2>
    Search people
  </styled.h2>
  <styled.div>
    <SearchBox
      disabled={false}
      id="searchField"
      label="Search people"
      name="searchField"
      onSearchChange={[Function]}
      value=""
    />
    <CustomButton
      collapse={false}
      onClick={[Function]}
      selected={false}
      size="small"
      type="button"
      variant="default"
    >
      Search
    </CustomButton>
  </styled.div>
  <styled.div>
    <styled.table>
      <styled.thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Gender
          </th>
          <th>
            Birth year
          </th>
          <th>
            Eye color
          </th>
          <th />
        </tr>
      </styled.thead>
      <styled.tbody>
        <tr>
          <td
            colSpan={5}
          >
            All characeters in this page are already saved
          </td>
        </tr>
      </styled.tbody>
    </styled.table>
  </styled.div>
  <styled.div>
    <Pagination />
  </styled.div>
</styled.div>
`;

exports[`<SearchPeople /> tests store with data tests matches snapshot 1`] = `
<styled.div>
  <styled.h2>
    Search people
  </styled.h2>
  <styled.div>
    <SearchBox
      disabled={false}
      id="searchField"
      label="Search people"
      name="searchField"
      onSearchChange={[Function]}
      value=""
    />
    <CustomButton
      collapse={false}
      onClick={[Function]}
      selected={false}
      size="medium"
      type="button"
      variant="default"
    >
      Search
    </CustomButton>
  </styled.div>
  <styled.div>
    <styled.table>
      <styled.thead>
        <tr>
          <th>
            Name
          </th>
          <th>
            Gender
          </th>
          <th />
        </tr>
      </styled.thead>
      <styled.tbody>
        <Character
          character={
            Object {
              "birthYear": "112BBY",
              "eyeColor": "yellow",
              "gender": "n/a",
              "hairColor": "n/a",
              "height": "167",
              "mass": "75",
              "name": "C-3PO",
              "skinColor": "gold",
            }
          }
          key="C-3PO"
          viewWidth={500}
        />
        <Character
          character={
            Object {
              "birthYear": "41.9BBY",
              "eyeColor": "yellow",
              "gender": "male",
              "hairColor": "none",
              "height": "202",
              "mass": "136",
              "name": "Darth Vader",
              "skinColor": "white",
            }
          }
          key="Darth Vader"
          viewWidth={500}
        />
        <Character
          character={
            Object {
              "birthYear": "52BBY",
              "eyeColor": "blue",
              "gender": "male",
              "hairColor": "brown, grey",
              "height": "178",
              "mass": "120",
              "name": "Owen Lars",
              "skinColor": "light",
            }
          }
          key="Owen Lars"
          viewWidth={500}
        />
        <Character
          character={
            Object {
              "birthYear": "47BBY",
              "eyeColor": "blue",
              "gender": "female",
              "hairColor": "brown",
              "height": "165",
              "mass": "75",
              "name": "Beru Whitesun lars",
              "skinColor": "light",
            }
          }
          key="Beru Whitesun lars"
          viewWidth={500}
        />
        <Character
          character={
            Object {
              "birthYear": "unknown",
              "eyeColor": "red",
              "gender": "n/a",
              "hairColor": "n/a",
              "height": "97",
              "mass": "32",
              "name": "R5-D4",
              "skinColor": "white, red",
            }
          }
          key="R5-D4"
          viewWidth={500}
        />
        <Character
          character={
            Object {
              "birthYear": "24BBY",
              "eyeColor": "brown",
              "gender": "male",
              "hairColor": "black",
              "height": "183",
              "mass": "84",
              "name": "Biggs Darklighter",
              "skinColor": "light",
            }
          }
          key="Biggs Darklighter"
          viewWidth={500}
        />
        <Character
          character={
            Object {
              "birthYear": "57BBY",
              "eyeColor": "blue-gray",
              "gender": "male",
              "hairColor": "auburn, white",
              "height": "182",
              "mass": "77",
              "name": "Obi-Wan Kenobi",
              "skinColor": "fair",
            }
          }
          key="Obi-Wan Kenobi"
          viewWidth={500}
        />
      </styled.tbody>
    </styled.table>
  </styled.div>
  <styled.div>
    <Pagination />
  </styled.div>
</styled.div>
`;
